### PR TITLE
Add `scale-writers.max-nodes-count` property to limit writer scaling

### PR DESCRIPTION
**Pull request by radek-starburst**
## Description 

Added option `task.scale-writers.max-nodes-count` to `TaskManagerConfig` and a session option to make it possible to limit writer scaling by number of nodes that can be used by `ScaledWriterScheduler`.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
(x ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java [hunk 1]
```diff
@@ -78,6 +78,7 @@
     private Duration interruptStuckSplitTasksDetectionInterval = new Duration(2, TimeUnit.MINUTES);
 
     private boolean scaleWritersEnabled = true;
+    private int scaleWritersMaxNodesCount = 100;
```
Comment (gaurav8297) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:81]: Why 100?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java [hunk 2]
```diff
@@ -109,15 +130,12 @@ private int getNewTaskCount()
...
 
-        // When there is a big data skewness, there could be a bottleneck due to the skewed workers even if most of the workers are not over-utilized.
-        // Check both, weighted output buffer over-utilization rate and average output buffer over-utilization rate, in case when there are many over-utilized small tasks
-        // due to fewer not-over-utilized big skewed tasks.
-        if ((isWeightedBufferFull() || isAverageBufferFull()) && (writtenBytes >= minWrittenBytesToScaleUp)) {
-            return 1;
-        }
-
-        return 0;
+    private boolean isUpperLimitExceeded()
```
Comment (gaurav8297) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java:136]: Do we need to have a separate method? I think it's better to have this condition directly in the above if condition

File: core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java [hunk 3]
```diff
@@ -148,10 +148,32 @@ public void testGetNewTaskCountWhenExistingWriterTaskMaxWriterCountIsEmpty()
...
+        TaskStatus taskStatus = buildTaskStatus(true, 123456L);
+        AtomicReference<List<TaskStatus>> taskStatusProvider = new AtomicReference<>(ImmutableList.of(taskStatus));
+        ScaledWriterScheduler scaledWriterScheduler = buildScaledWriterScheduler(taskStatusProvider, 2);
+
+        scaledWriterScheduler.schedule();
+        assertEquals(scaledWriterScheduler.schedule().getNewTasks().size(), 1);
+    }
+
+    @Test
+    public void testNewTaskCountWhenUpperLimitIsExceeded()
```
Comment (gaurav8297) [inline core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java:163]: `testNewTaskCountWhenUpperLimitIsExceeded` -> `testNewTaskCountWhenNodesUpperLimitIsExceeded`

**Review by gaurav8297 — COMMENTED** [review]
Mostly LGTM! There are some tests failing

File: core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java (hunk 2 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java:136]: ok

File: core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java (hunk 3 above)
Comment (radek-starburst) [reply core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java:163]: ok

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java (hunk 1 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:81]: Default value for `hashPartitionCount` is `100`. These options in some sense similar (in the sense of introducing a limit on number of nodes taking part in the query's stages). We need to decide on any value, the `100` seems to be reasonable choice.

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java [hunk 4]
```diff
@@ -447,6 +448,21 @@ public TaskManagerConfig setScaleWritersMaxWriterCount(int scaleWritersMaxWriter
...
 
+    @Min(1)
+    public int getScaleWritersMaxNodesCount()
+    {
+        return scaleWritersMaxNodesCount;
+    }
+
+    @Config("task.scale-writers.max-nodes-count")
+    @ConfigDescription("Maximum number of nodes that will take part in writer tasks. It is an upper bound on scaling of writers " +
+            "and works only if task.scale-writers.enabled is set")
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:459]: Node limit should also apply `task.scale-writers.enabled` when `preferred partitioning` insert mode is used. Please refer to @gaurav8297 for help

File: core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java [hunk 5]
```diff
@@ -447,6 +448,21 @@ public TaskManagerConfig setScaleWritersMaxWriterCount(int scaleWritersMaxWriter
         return this;
     }
 
+    @Min(1)
+    public int getScaleWritersMaxNodesCount()
+    {
+        return scaleWritersMaxNodesCount;
+    }
+
+    @Config("task.scale-writers.max-nodes-count")
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:457]: This should be moved to `QueryManagerConfig` below `query.hash-partition-count` setting.
Name it `query.max-writer-node-count` or something similar.

**Comment by sopel39** [comment]
@radek-starburst  please rebase

**Pull request closed without merging**